src/checkers/inference/model/ConstraintManager.java

```diff
@@ -167,41 +144,92 @@ private AnnotationLocation getCurrentLocation() {
-        } else {
-            this.add(this.createSubtypeConstraint(subtype, supertype));
+        Constraint constraint = createSubtypeConstraint(subtype, supertype);
+
+        if (constraint instanceof AlwaysFalseConstraint) {
+            checker.report(Result.failure("subtype.constraint.unsatisfiable", subtype, supertype),
+                    visitorState.getPath().getLeaf());
         }
+
+        add(constraint);
     }
 
+    /**
+     * Creates and adds an {@link EqualityConstraint} between the two slots to the constraint set,
+     * which may be normalized to {@link AlwaysTrueConstraint}. An error is issued if the equality
+     * constraint is always unsatisfiable.
+     */
     public void addEqualityConstraint(Slot first, Slot second) {
-        this.add(this.createEqualityConstraint(first, second));
+        Constraint constraint = createEqualityConstraint(first, second);
+
+        if (constraint instanceof AlwaysFalseConstraint) {
+            checker.report(Result.failure("equality.constraint.unsatisfiable", first, second),
+                    visitorState.getPath().getLeaf());
+        }
+
+        add(constraint);
     }
 
+    /**
+     * Creates and adds an {@link InequalityConstraint} between the two slots to the constraint set,
+     * which may be normalized to {@link AlwaysTrueConstraint}. An error is issued if the inequality
+     * constraint is always unsatisfiable.
+     */
     public void addInequalityConstraint(Slot first, Slot second) {
-        this.add(this.createInequalityConstraint(first, second));
+        Constraint constraint = createInequalityConstraint(first, second);
+
+        if (constraint instanceof AlwaysFalseConstraint) {
+            checker.report(Result.failure("inequality.constraint.unsatisfiable", first, second),
+                    visitorState.getPath().getLeaf());
+        }
+
+        add(constraint);
     }
 
+    /**
+     * Creates and adds a {@link ComparableConstraint} between the two slots to the constraint set,
+     * which may be normalized to {@link AlwaysTrueConstraint}. An error is issued if the comparable
+     * constraint is always unsatisfiable.
+     */
     public void addComparableConstraint(Slot first, Slot second) {
-        this.add(this.createComparableConstraint(first, second));
+        Constraint constraint = createComparableConstraint(first, second);
+
+        if (constraint instanceof AlwaysFalseConstraint) {
+            checker.report(Result.failure("comparable.constraint.unsatisfiable", first, second),
+                    visitorState.getPath().getLeaf());
+        }
+
+        add(constraint);
     }
 
+    /**
+     * Creates and adds a {@link CombineConstraint} to the constraint set.
+     */
     public void addCombineConstraint(Slot target, Slot decl, Slot result) {
-        this.add(this.createCombineConstraint(target, decl, result));
+        add(createCombineConstraint(target, decl, result));
```

Comment:
add check for False here too, and below. Keep return of Create the specific Constraint type. If future PR changes return type, the check here stays.